torch/csrc/jit/autodiff.cpp

```diff
@@ -230,30 +236,35 @@ static value_set findAllRequiresGradNodes(
   return requires_grad_set;
 }
 
-static Value* createZerosLike(Value *v) {
-  JIT_EXPECTM(v->isTensor(), "can't allocate zero gradient for a value without a type");
-  Graph *graph = v->owningGraph();
-  auto type = v->type()->expect<TensorType>();
-  at::DeviceGuard device_guard(type->device());
-
-  auto & at_type = type->device() == -1 ? at::CPU(type->scalarType()) : at::CUDA(type->scalarType());
-  auto zeros = at::zeros({}, at_type).expand(type->sizes());
-  Node *constant = graph->createConstant(zeros)
-                        ->i_(attr::is_zero, 1);
-  graph->insertNode(constant);
-  return constant->output();
-}
 
-// any vjp input may be undefined, and we need to potentially replace it
-// with a zero tensor of the right size if required.
-// this function inserts a guard into the graph that does this replacement.
-// ReplaceIfUndef(dv,c) replaces dv with c if dv is undef.
-// During Graph specialization these guards will get removed when
-// 'dv' is known to be undef, and the zeros will be propagated if possible.
-static Value* createUndefGuard(Value * dv, Value * alternative) {
-  Graph* graph = dv->owningGraph();
-  Node * n = graph->create(prim::ReplaceIfUndef, {dv, alternative});
-  return graph->insertNode(n)->output();
+// If we have a function y = f(x) with jacobian J, the backwards of f is dx = J^t dy.
+// Note that because the backwards always implements this matrix multiply,
+// we know that it maps an input vector of zeros to an output vector of zero
+// regardless of what operations it choses to do inside to actually implement
+// the matrix multiply (most use some optimized form and never generate J^t).
+// More generally, we know that all of the backward computations are linear and
+// can use this property to do more aggressive optimizations later.
+// It is ok to replace any backward function with known-zero inputs with something
+// that produces known-zero outputs. This function encloses each know-linear
+// backward function in a 'LinearBlock' sub-block so that we can perform optimizations
+// using this information. In particular, specializeUndef will observe if
+// all the inputs to the linear block are Undef, which the autograd uses to represent
+// zeros, and then propagate the undefs to the outputs of the block.
+static std::vector<Value*> linearGradientForNode(Node* node, ArrayRef<Value*> grad_values) {
+  auto & graph = *node->owningGraph();
+  auto linear = graph.insertNode(graph.create(prim::LinearBlock, {grad_values}, 0));
+  // to make reading gradient graphs easier, remember the name of the forward op
```

Comment:
Can we call this `backward_of`? I read the graphs and I thought, "weird, this says it's an add, but I don't see any add inside the block" until I got to this comment.

Reply:
+1 for `backward_of` or `grad_of`, which takes up less space